### ISSUE TITLE
`optimizer.zero_grad` called after calculating loss instead of before in `lecture3_flow_models_demos.ipynb`.

### ISSUE THREAD

**Issue by Kinyugo** [rll/deepul#12]
In the **lecture3_flow_models_demos.ipynb** optimizer.zero_grad` is called after calculating the loss. i.e: 
```python
 model.train()
  for x in train_loader:
      x = x.float()
      loss = model.nll(x)
      optimizer.zero_grad()
      loss.backward()
      optimizer.step()
```
In other implementation, the standard is to call `optimizer.zero_grad` before forward propagation. e.g
```python
optimizer.zero_grad()
y = model(x)
loss = criterion(y, y_true)
loss.backward()
optimizer.step()
```
Changing the implementation to before calculating the loss, breaks the model and the model fails to learn anything. I cannot seem to understand why the model has that behavior.
